### DISCUSSION TITLE
A Littlle Bug in Sim-OTA

### DISCUSSION THREAD

**Discussion by kellenf** [open-mmlab/mmdetection#9095] [Q&A] [answered]
I found a little bug in sim-ota code
`def dynamic_k_matching(self, cost, pairwise_ious, num_gt, valid_mask):[1 year ago • RangiLyu [[Feature] Add YOLOXHead, SimOTAAssigner, and …]](https://sourcegraph.com/github.com/open-mmlab/mmdetection/-/commit/ec76658a78df9d2e14c691d320e78e84e7464f98)
        matching_matrix = torch.zeros_like(cost, dtype=torch.uint8)
        # select candidate topk ious for dynamic-k calculation
        candidate_topk = min(self.candidate_topk, pairwise_ious.size(0))
        topk_ious, _ = torch.topk(pairwise_ious, candidate_topk, dim=0)
        # calculate dynamic k for each gt
        dynamic_ks = torch.clamp(topk_ious.sum(0).int(), min=1)
        for gt_idx in range(num_gt):
            _, pos_idx = torch.topk(
                cost[:, gt_idx], k=dynamic_ks[gt_idx], largest=False)
            matching_matrix[:, gt_idx][pos_idx] = 1

        del topk_ious, dynamic_ks, pos_idx

        prior_match_gt_mask = matching_matrix.sum(1) > 1
        if prior_match_gt_mask.sum() > 0:
           # if all dynamic_k == 1, there is probility one point is assigned to multi gt-box, but use this code, it will cause some gt is assigned no point???
            cost_min, cost_argmin = torch.min(
                cost[prior_match_gt_mask, :], dim=1)
            matching_matrix[prior_match_gt_mask, :] *= 0
            matching_matrix[prior_match_gt_mask, cost_argmin] = 1
        # get foreground mask inside box and center prior
        fg_mask_inboxes = matching_matrix.sum(1) > 0
        valid_mask[valid_mask.clone()] = fg_mask_inboxes

        matched_gt_inds = matching_matrix[fg_mask_inboxes, :].argmax(1)
        matched_pred_ious = (matching_matrix *
                             pairwise_ious).sum(1)[fg_mask_inboxes]
        return matched_pred_ious, matched_gt_inds`
In sim-OTA, if dynamic_k == 1, it will cause no point is matched to some gt-box in some probility? Why use this

**Comment by RangiLyu** [comment, marked as answer]
We followed the official implementation of YOLOX. And actually, one point should only match one gt, otherwise, it will cause ambiguity in the regression branch.
